beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -3117,19 +3140,83 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
     /// The provided `state_root_opt` should only ever be set to `Some` if the contained value is
     /// equal to the root of `state`. Providing this value will serve as an optimization to avoid
     /// performing a tree hash in some scenarios.
-    pub fn produce_block_on_state<Payload: ExecPayload<T::EthSpec>>(
-        &self,
-        mut state: BeaconState<T::EthSpec>,
+    pub async fn produce_block_on_state<Payload: ExecPayload<T::EthSpec>>(
+        self: &Arc<Self>,
+        state: BeaconState<T::EthSpec>,
         state_root_opt: Option<Hash256>,
         produce_at_slot: Slot,
         randao_reveal: Signature,
         validator_graffiti: Option<Graffiti>,
         verification: ProduceBlockVerification,
     ) -> Result<BeaconBlockAndState<T::EthSpec, Payload>, BlockProductionError> {
-        let eth1_chain = self
-            .eth1_chain
-            .as_ref()
-            .ok_or(BlockProductionError::NoEth1ChainConnection)?;
+        // Part 1/3 (blocking)
+        //
+        // Perform the state advance and block-packing functions.
+        let chain = self.clone();
+        let mut partial_beacon_block = self
+            .task_executor
+            .spawn_blocking_handle(
```

Comment:
another opportunity to simplify to `self.spawn_blocking_handle`

Reply:
As above.